sdk/servicebus/azure_messaging_servicebus/src/common/authorizer.rs

```diff
@@ -0,0 +1,487 @@
+// Copyright (c) Microsoft Corporation. All Rights reserved
+// Licensed under the MIT license.
+
+use async_lock::Mutex as AsyncMutex;
+use azure_core::{
+    async_runtime::{get_async_runtime, SpawnedTask},
+    credentials::{AccessToken, TokenCredential},
+    error::ErrorKind as AzureErrorKind,
+    fmt::SafeDebug,
+    http::Url,
+    time::{Duration, OffsetDateTime},
+    Result as AzureResult,
+};
+use azure_core_amqp::{AmqpClaimsBasedSecurityApis as _, AmqpConnection, AmqpSessionApis as _};
+use rand::{rng, Rng};
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex as SyncMutex, OnceLock, Weak},
+};
+use tracing::{debug, trace, warn};
+
+// The number of seconds before token expiration that we wake up to refresh the token.
+const TOKEN_REFRESH_BIAS: Duration = Duration::minutes(6); // By default, we refresh tokens 6 minutes before they expire.
+const TOKEN_REFRESH_JITTER_MIN: Duration = Duration::seconds(-5); // Minimum jitter (added from the bias, so a negative number means we refresh before the bias)
+const TOKEN_REFRESH_JITTER_MAX: Duration = Duration::seconds(5); // Maximum jitter (added to the bias)
```

Comment:
These numbers may be a bit tight - worst case, you're going to refresh 1 minute before the token expires.